File: Completed/Go/257.go
import (
	"strconv"
)

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func binaryTreePaths(root *TreeNode) []string {
	results := make([]string, 0)
	if root == nil {
		return results
	}

	initalPath := strconv.Itoa(root.Val)
	if root.Left == nil && root.Right == nil {
		results = append(results, initalPath)
	} else {
		helper(root.Left, initalPath, &results)
		helper(root.Right, initalPath, &results)
	}

	return results
}

func helper(root *TreeNode, path string, rootToLeaf *[]string) {
	if root == nil {
		return
	}

	newPath := path + "->" + strconv.Itoa(root.Val)
	if root.Left == nil && root.Right == nil {
		*rootToLeaf = append(*rootToLeaf, newPath)
	}

	helper(root.Left, newPath, rootToLeaf)
	helper(root.Right, newPath, rootToLeaf)
}

File: Completed/python3/257.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: TreeNode) -> List[str]:
      result = []
      if root == None:
        return result
      
      path = str(root.val)
      if root.left == None and root.right == None:
        return [path]
      
      self.dfs(root.left, path, result)
      self.dfs(root.right, path, result)
      
      return result
    
    def dfs(self, root: TreeNode, path: str, result: List[str]):
      if root == None:
        return
      
      path += "->" + str(root.val)
      if root.left == None and root.right == None:
        result.append(path)
        return
      
      self.dfs(root.left, path, result)
      self.dfs(root.right, path, result)
        
